src/osd/ReplicatedPG.cc

```diff
@@ -1496,6 +1496,22 @@ void ReplicatedPG::do_op(OpRequestRef& op)
     return;
   }
 
+  // discard due to cluster full transition?  (we discard any op that
+  // originates before the cluster or pool is marked full; the client
+  // will resend after the full flag is removed or if they expect the
+  // op to succeed despite being full).
+  // FIXME: we exclude mds writes for now.
+  if (!m->get_source().is_mds() &&
+      info.history.last_epoch_marked_full > m->get_map_epoch()) {
+    dout(10) << __func__ << " discarding op sent before full " << m << " "
+	     << *m << dendl;
+    return;
+  }
+  if (osd->check_failsafe_full()) {
+    osd->reply_op_error(op, -ENOSPC);
```

Comment:
> The reason I'm worried about returning ENOSPC is that rbd clients can't handle it (http://tracker.ceph.com/issues/6938) -
> that's why we removed it in the past. It could cause corruption on krbd, since the block layer just translates any error
> to EIO.
> 
> We could make the client retry on ENOSPC (maybe only for clients that want that handled for them, like rbd), but I'd worry
> about old clients without ENOSPC handling (esp. krbd).

Ah, right.  Okay, maybe we should only send ENOSPC if the client indicates 
it is knowingly sending a request despite FULL.

> I was misreading the failsafe logic - it's just setting the local full state and not updating the monitors at the same
> time. Also there's no protection for the osd failsafe thresholds being lower than the monitor thresholds, so a
> misconfiguration could cause dropped writes with the client never seeing any full flags. It seems like the osd failsafes
> should be max(osd_failsafe, mon_setting), and we should update the mon when the osd reaches it to be sure the cluster full
> flag gets set. Then we can be sure clients will resend in the future, since they will see an epoch where the cluster was
> marked full.
> 
> I don't see why deletes need to get an explicit error here - I thought the goal was to let them through all
> cluster/pool/osd failsafe space checks?

Hrm.. yeah.  It seems like the failsafe is also kinda broken.  Maybe it 
should

 - set a flag indicating it reached a full threshold
 - request a FULL state from the mon
 - drop all further requests

...and that state would get cleared if it was marked down.  The mon, if it 
got such a request, could either mark the cluster FULL, or (if the 
threshold is too low) mark the sender down.

That introduces the possibility that the osd could flap because it thinks 
it's too full and the mon disagrees, though.. but I'm not sure how to get 
around that type of misconfiguration.  Maybe if the mon's full thresholds 
are actually published in the osdmap instead of config options?